### PR TITLE
Use probes more aggressively in new solver

### PR DESCRIPTION
**Pull request by compiler-errors**
....so that we have the right candidate information when assembling trait and normalizes-to goals.

Also gets rid of misc probes.

r? lcnr

### CODE REVIEW

File: compiler/rustc_middle/src/traits/solve.rs [hunk 1]
```diff
@@ -332,4 +332,7 @@ pub enum CandidateSource {
     /// }
     /// ```
     AliasBound,
+    /// Uwu
```
Comment (lqd) [inline compiler/rustc_middle/src/traits/solve.rs:335]: TwT

Comment (compiler-errors) [reply compiler/rustc_middle/src/traits/solve.rs:335]: oh lol

File: compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs [hunk 2]
```diff
@@ -517,104 +518,112 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
...
-                ty::Error(e) => Ty::new_error(tcx, *e),
-
-                ty::Str | ty::Slice(_) => tcx.types.usize,
-
-                ty::Dynamic(_, _, ty::Dyn) => {
-                    let dyn_metadata = tcx.require_lang_item(LangItem::DynMetadata, None);
-                    tcx.type_of(dyn_metadata)
-                        .instantiate(tcx, &[ty::GenericArg::from(goal.predicate.self_ty())])
-                }
+        ecx.probe_trait_candidate(CandidateSource::BuiltinImpl(BuiltinImplSource::Misc)).enter(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs:543]: maybe add `probe_builtin_trait_candidate` and use 
```suggestion
        ecx.probe_builtin_trait_candidate(BuiltinImplSource::Misc).enter(|ecx| {
```
unhappy about formatting here :3

File: compiler/rustc_trait_selection/src/traits/coherence.rs [hunk 3]
```diff
@@ -1079,8 +1079,8 @@ impl<'a, 'tcx> ProofTreeVisitor<'tcx> for AmbiguityCausesVisitor<'a, 'tcx> {
         let mut ambiguity_cause = None;
         for cand in goal.candidates() {
             // FIXME: boiiii, using string comparisions here sure is scuffed.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/coherence.rs:1081]: :<

**Review by lcnr — COMMENTED** [review]
happy with this PR, you have to fix how `analyse.rs` collects candidates though, maybe rely on #124444

**Comment by lcnr** [comment]
r=me after removing the "lol" comment :3

**Pull request merged**